File: yandex/lec5/3/3.py
def get_prefix(array):
    prefix = [0] * (len(array) + 1)
    for i in range(1, len(array)):
        h = array[i] - array[i - 1]
        if h > 0:
            prefix[i + 1] = prefix[i] + h
        else:
            prefix[i + 1] = prefix[i]
    return prefix


if __name__ == '__main__':
    n = int(input())
    coords = [int(input().split()[1]) for i in range(n)]
    m = int(input())
    ran = [list(map(int, input().split())) for i in range(m)]

    prefix_count = get_prefix(coords)
    reverse_prefix = get_prefix(coords[::-1])

    for item in ran:
        l, r = item
        print(prefix_count)
        print(reverse_prefix)
        if l < r:
            count = prefix_count[r] - prefix_count[l]
        else:
            count = reverse_prefix[l] - reverse_prefix[r + 1]
        print(count)

File: yandex/lec7/3/3.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

const (
	CAN  = 0
	CANT = 1
)

func main() {
	sc := bufio.NewScanner(os.Stdin)
	nd := toInt(sc)
	n, d := nd[0], nd[1]
	coords := toInt(sc)
	var events [][]int
	for i := 0; i < n; i++ {
		events = append(events, []int{coords[i], CAN})
		events = append(events, []int{coords[i] + d, CANT})
	}
	sort.Slice(events, func(i, j int) bool {
		return events[i][0] < events[j][0] ||
			(events[i][0] == events[j][0] && events[i][1] < events[j][1])
	})
	max, count := 0, 0

	for _, event := range events {
		fmt.Println("count:", count)
		eventType := event[1]
		if eventType == CAN {
			count++
		} else if eventType == CANT {
			count--
		}

		if count > max {
			max = count
		}
	}
	dict := map[int]int{}
	variant := 1
	for _, event := range events {
		x, evenType := event[0], event[1]
		if evenType == CAN {
			dict[x] = variant
		}
		variant++
		if variant > max {
			variant = 1
		}
	}

	var res []string
	for _, point := range coords {
		res = append(res, strconv.Itoa(dict[point]))
	}

	fmt.Println(max)
	fmt.Println(strings.Join(res, " "))

}

func toInt(scanner *bufio.Scanner) []int {
	scanner.Scan()
	array := strings.Split(scanner.Text(), " ")
	var res []int
	for _, item := range array {
		intItem, _ := strconv.Atoi(item)
		res = append(res, intItem)
	}
	return res
}
